### PR TITLE
changes to make xt_nlp_annotator work

### PR DESCRIPTION
**Pull request by abishekarun**
<summary of the change, relevant motivation, and context>

Change type:

- [ ] Bug fix
- [x] Feature
- [ ] Documentation

Checklist:

- [x] My code follows the style guidelines of this [project](https://docs.google.com/document/d/1jckZJe0CrWyF-IjxoO2OfBKAyKLC3Yk9Xr7UmOLBETA/edit?usp=sharing)
- [x] I have performed a self-review of my own code
- [x] I have commented my code and added docstrings to all exposed functions and classes
- [x] I have made corresponding changes to the documentation
- [x] Any relevant changes to third party licenses have been updated in the README

### CODE REVIEW

File: xt_training/runner.py [hunk 1]
```diff
@@ -1,10 +1,13 @@
 import torch
 import numpy as np
 import json
-from collections import Iterable
+import logging
```
Comment (timesler) [inline xt_training/runner.py]: This isn't used is it?

File: xt_training/runner.py [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
 
+from collections import Iterable
 from .metrics import EPS, PooledMean, Metric
 
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
```
Comment (timesler) [inline xt_training/runner.py]: This shouldn't be here should it?

File: xt_training/runner.py [hunk 3]
```diff
@@ -28,6 +31,20 @@ def __call__(self, loss, metrics, i):
             print('')
 
 
+def detach_objects(x):
```
Comment (timesler) [inline xt_training/runner.py:30]: Can you add a docstring here?

File: xt_training/runner.py [hunk 4]
```diff
@@ -28,6 +31,20 @@ def __call__(self, loss, metrics, i):
...
+        out = []
+        for x_i in x:
+            if isinstance(x_i, torch.Tensor):
+                out.append(x_i.detach().cpu())
+            else:
+                out.append(x_i)
+        return out
+    else:
+        return x
+
```
Comment (timesler) [inline xt_training/runner.py:54]: 2 lines between top level classes and functions, 1 line between method functions

File: xt_training/runner.py [hunk 5]
```diff
@@ -145,10 +176,10 @@ def __call__(self, loader, mode=None, return_preds=False):
...
                 if return_preds:
-                    y_pred_epoch.append(y_pred.detach().cpu())
-                    y_epoch.append(y.detach().cpu())
+                    y_pred_epoch.append(detach_objects(y_pred))
+                    y_epoch.append(detach_objects(y))
 
-        if model.training and scheduler is not None:
+        if model.training and scheduler is not None and not is_batch_scheduler:
             scheduler.step()
             self.epoch += 1
```
Comment (timesler) [inline xt_training/runner.py]: This will break the self.epoch counter. I think it should be:
```python
if model.training:
    if not is_batch_scheduler and scheduler is not None:
        scheduler.step()
    self.epoch += 1
```

(This was already broken by me the way it was, don't feel bad :))

**Review by timesler — CHANGES_REQUESTED** [review]

File: xt_training/runner.py (hunk 1 above)
Comment (abishekarun) [reply xt_training/runner.py]: It is used to check if x and y are lists. Also, in the detach objects function where we need to check if it is a list of tensors.

Comment (DavidBegert) [reply xt_training/runner.py]: I assume he was talking about logging hahah

File: xt_training/runner.py [hunk 6]
```diff
@@ -145,10 +185,11 @@ def __call__(self, loader, mode=None, return_preds=False):
...
                 if return_preds:
-                    y_pred_epoch.append(y_pred.detach().cpu())
-                    y_epoch.append(y.detach().cpu())
+                    y_pred_epoch.append(detach_objects(y_pred))
+                    y_epoch.append(detach_objects(y))
 
-        if model.training and scheduler is not None:
+        if model.training:
+            if not is_batch_scheduler and scheduler is not None:
             scheduler.step()
```
Comment (DavidBegert) [inline xt_training/runner.py]: indentation

**Review by DavidBegert — APPROVED** [review]
